src/MagicWormhole/Internal/Rendezvous.hs

```diff
@@ -65,44 +61,90 @@ import MagicWormhole.Internal.WebSockets (WebSocketEndpoint(..))
+data Session
+  = Session
+  { pendingVar :: TVar (HashMap ResponseType (TMVar Messages.ServerMessage))
+  , messageChan :: TQueue Messages.MailboxMessage
+  , motd :: TMVar (Maybe Text)
+  , connection :: WS.Connection
+  }
+
+-- | Create a new 'Session'.
+new :: WS.Connection -- ^ Active WebSocket connection to a Rendezvous Server.
+    -> STM Session  -- ^ Opaque 'Session' object.
+new connection
+  = Session
+  <$> newTVar mempty
+  <*> newTQueue
+  <*> newEmptyTMVar
+  <*> pure connection
+
+-- | Send a message to a Magic Wormhole Rendezvous server.
+send :: Session -- ^ An active session. Get this using 'with'.
+     -> Messages.ClientMessage -- ^ Message to send to the server.
+     -> IO ()
+-- XXX: I think this needs to use `sendClose` when the message is Close.
+send session msg = do
+  WS.sendBinaryData (connection session) (encode msg)
+  putStrLn @Text $ ">>> " <> show msg -- XXX: Debug
+
+-- | Receive a message from a Magic Wormhole Rendezvous server.
+-- Blocks until such a message exists.
+receive :: Session -- ^ An active session. Get this using 'with'.
+        -> IO (Either ServerError Messages.ServerMessage) -- ^ Next message from the server.
+receive session = do
+  -- XXX: I think we need to catch `CloseRequest` here and gracefully stop.
+  msg <- WS.receiveData (connection session)
+  case eitherDecode msg of
+    Left err -> do
+      putStrLn @Text $ "[ERROR] " <> show err
+      pure $ Left (ParseError err)
+    Right result -> do
+      putStrLn @Text $ "<<< " <> show result  -- XXX: Debug
+      pure $ Right result
+
+-- | Run a Magic Wormhole Rendezvous client. Use this to interact with a Magic Wormhole server.
 --
 -- Will fail with IO (Left ServerError) if the server declares we are unwelcome.
-runClient :: HasCallStack => WebSocketEndpoint -> Messages.AppID -> Messages.Side -> (Session -> IO a) -> IO (Either ServerError a)
-runClient (WebSocketEndpoint host port path) appID side' app = do
-  inputChan <- atomically newTChan
-  outputChan <- atomically newTChan
+runClient
+  :: HasCallStack
+  => WebSocketEndpoint -- ^ The websocket to connect to
+  -> Messages.AppID -- ^ ID for your application (e.g. example.com/your-application)
+  -> Messages.Side -- ^ Identifier for your side
+  -> (Session -> IO a) -- ^ Action to perform inside the Magic Wormhole session
+  -> IO (Either ServerError a) -- ^ The result of the action or a ServerError
+runClient (WebSocketEndpoint host port path) appID side' app =
   map join $ Socket.withSocketsDo . WS.runClient host port path $ \ws -> do
-    result <- race (race (socketToChan ws inputChan) (chanToSocket ws outputChan))
-                   (with inputChan outputChan action)
-    pure $ case result of
-             Left (Left readErr) -> Left readErr
-             Left (Right writeErr) -> Left writeErr
-             Right result' -> result'
+    session <- atomically $ new ws
+    race (readMessages session) (action session)
```

Comment:
I don't think I understand what's going on here.   I see that `race` runs two IO monads concurrently, completing with the result of the first to complete and cancelling the other.  I see that `readMessages` read all message and dispatches them to some application logic.  I see that `action` binds the connection to an appID and a side.

I don't see why you would want these two things to race rather than just doing one and then the other.

Reply:
`readMessages` is an infinite loop, reading messages from the websocket all the time. It will only terminate on an error.

`action` does two things:
  1. bind a session to an appID and side
  2. and then runs the user-provided `app`, passing it the `session` we created

We don't need to be running `readMessages` during the `bind` call (although, why not?), we do need to be running it while `app` is doing its thing, so that requests get responses.

Does this make sense?